### ISSUE TITLE
C++14/C++17 in LAMMPS

### ISSUE THREAD

**Issue by rbberger** [lammps/lammps#3840]
## Summary

It's been 5 years since issue #1659 (2019), where we kept track of the C++11 transition for LAMMPS, and 4 years (2020) since C++11 was allowed in LAMMPS core.

In a similar way, this issue now documents the current state of common compilers among popular Linux distributions.
By now, moving to C++14 is a trivial change, since most compilers at least default to C++14, if not already C++17.

The only exception is the default RHEL7 compiler, which barely supports C+11. I personally don't consider this a blocker, given that most HPC centers, even if they still use RHEL7-based systems, will provide newer GCC compilers.

As a side note, GCC versions prior to v9.x should not be used for C++17 due to various quirks and bugs.

Currently, C++17 is currently only required by Kokkos 4.

C++20 will be required in Kokkos 5.

## Proposal to assess impact
```c++
// for C++14
#ifndef LAMMPS_CXX14
#if __cplusplus <= 201402L
  #error LAMMPS is planning to transition to C++14. Do disable this error please use a C++14-compliant compiler or define LAMMPS_CXX14 in your makefile
#endif
#endif

// alternative for C++17
#ifndef LAMMPS_CXX17
#if __cplusplus <= 201703L
  #error LAMMPS is planning to transition to C++17. Do disable this error please use a C++17-compliant compiler or define LAMMPS_CXX17 in your makefile
#endif
#endif
```

## Target Platforms

| OS |  g++ Version | g++ default | clang++ Version | clang++ default |
| --- | ---- | --- | --- | ---|
| Ubuntu 18.04 LTS   | 7.4.0 | C++14    | 6.0.0 | C++14 |
| Ubuntu 20.04 LTS (EOS 04/2023)   |  9.4.0  | C++14  | 10.0.0 | C++14 |
| Ubuntu 22.04 LTS (EOS 04/2027)   |  11.3.0 | C++17  | 14.0.0 | C++14 |
| Ubuntu 24.04 LTS              |  13.2.0 | C++17  | 18.1.3 | C++17 |
| RHEL7 (EOL July 2024)    | 4.8.5 | C++98 (up to C++11 with `-std=c++11`) | 3.4.2 | C++98 (up to C++11 with `-std=c++11`) |
| RHEL 8    | 8.5.0 | C++14 | 15.0.7 | C++14 |
| RHEL 9    | 11.3.1 | C++17 | 15.0.7 | C++14 |
| Fedora 37 (EOL 12/15/2023)     | 12.2.0 | C++14    | 15.0.7 | C++14 |
| Fedora 38     | 13.0.1 | C++17    | 16.0.5 | C++17 |
| Fedora 40     | 14.2.1 | C++17    | 18.1.6 | C++17 |
| OpenSuse Leap 15 | 7.3.9 | C++14    | 15.0.7 | C++14 |
| OpenSuse Tumbleweed | 13.1.1 | C++17    | 16.0.6 | C++17 |

### Testing default C++ compiler standard
```
g++ -dM -E -x c++  /dev/null | grep -F __cplusplus
```

## References:
 - https://en.cppreference.com/w/cpp/compiler_support

**Comment by akohlmey** [comment]
Unlike with C++11 there is currently little external pressure from contributors to use a more recent C++ standard. 

Internally, there is currently only Kokkos *requiring* a post-C++11 standard.

The latest Googletest release (1.15) requires C++14, but we don't use any of its features, so we can stick with version 1.12 for now. It also would not be a problem to raise the C++ standard only for testing, since this is not considered a core functionality of LAMMPS from the user perspective (very few users enable and use testing).

libfmt is "self-adapting" for C++11 to C++20. The latest development version still is compatible with C++11 without loss of functionality.

If we bring this to a LAMMPS developer discussion, there should be a list of benefits, i.e. new features or corrections in C++14 (or C++17?) that would have a positive impact on LAMMPS or could remove workarounds and/or inefficient or complex code. I would thus want to wait before adding the pre-processor check until a consensus about if and when to make a change is found.

**Comment by akohlmey** [comment]
This should be discussed at the Wednesday September 9th meeting.
Prerequisite is an update of the container images used in automatic testing.
We currently use CentOS 7, Ubuntu 20.04 LTS and Fedora 34. All of those are no longer supported.
Switching the baseline to Rocky/AlmaLinux 8, Ubuntu 22.04 LTS and Fedora 40/41 should allow to change the minimum requirements to: C++14 and CMake 3.20. This is still rather conservative compared to some other software (including Kokkos). If I understand correctly, the changes in C++17 and especially C++20 relative to the previous version are more drastic.

Major downside: cannot use default compilers on older OS versions.
Main benefit: use advanced features in fmtlib and googletest. benefit from improvements/corrections of C++14 over C++11.

**Comment by rbberger** [comment]
FWIW, here are my current thoughts about this.

There is nothing "drastic" in C++17 compared to C++14. There are mostly standard library improvements and bug fixes, which are all nice to have and a few enabling language features that primarily target actual C++ codes. I believe one jump in C++ standard to C++17 is easier than doing it in two jumps. The technical debt of moving to a newer standard isn't getting any less. Specifically since the actual significant upgrade would be C++20 which introduces C++ modules, `std::span`, `std::format`, and more `constexpr` in the standard library (which matters for GPU code). Since we are currently 4 standards behind, we could cut that in half with C++17.

That being said, one of the biggest blockers for C++17 has been that the IBM XL compilers only supports up to C++14 and are what is being used on Sierra. But even that constraint has been worked around by enabling mixed GCC and IBM XL builds.

Given that I stated that no GCC below 9.4 should be used for C++17, RHEL 8 would be a blocker.

Current Cray systems all use GCC 12+ / LLVM 17+ based compilers.

From the top of my head, after having worked on C++17 code bases for some time now, here's the stuff that I see most useful on a day-to-day basis (click the links for a full list for each standard).

[C++14 improvements](https://en.cppreference.com/w/cpp/14):
- mostly bug fixes compared to C++11
- generalized lambdas (you can use `auto` in more places)
- `constexpr` (aka make meta-programming readable)

[C++17 improvements](https://en.cppreference.com/w/cpp/17):
- `<filesystem>` library
- `std::variant`, `std::any`
- More attributes like `[[maybe_unused]]` or `[[nodiscard]]`
- More improved lambdas (`constexpr`)
- Fold expressions
- Structured bindings
- Inline variables
- Variable alias templates

For LAMMPS, none of these alone will be a silver bullet that magically transforms anything or break anything. But as we've seen with other refactors for C++11 so far, what we'll see over time is some code may get shorter, more readable or might even be removed and replaced by a standard library call. 

TL;DR:
- C++14: it's a no-op at this point. just do it.
- C++17: lots of nice to have, only blocker left are "old" Linux distributions on extended support

As for CMake, I believe that it should be its own issue. CMake 3.20 is a conservative minimum version, based on current extended support distros if we continue to care about them. But even with a minimum like that, we should test with newer CMake versions to detect breakage.

**Comment by rbberger** [comment]
The outcome of the LAMMPS Developer 2024 meeting was that we move to C++17 and CMake 3.20.

**Comment by junghans** [comment]
When we do the bump, we should consider to switch to use target based standards:
```
target_compile_features(lammps PUBLIC cxx_std_${LAMMPS_cxx_std})
```
to get the export targets right.
